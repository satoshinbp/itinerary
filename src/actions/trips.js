import db from '../firebase/firebase'

// ADD_TRIP
export const addTrip = trip => ({
  type: 'ADD_TRIP',
  trip
})

export const startAddTrip = (tripData = {}) => {
  return dispatch => {
    const {
      title = '',
      startDate = 0,
      endDate = 0,
      note = ''
    } = tripData
    const trip = { title, startDate, endDate, note }

    return db.collection('trips').add(trip).then(snapshot => {
      dispatch(addTrip({
        id: snapshot.id,
        ...trip
      }))
    })
  }
}

// EDIT_TRIP
export const editTrip = (id, updates) => ({
  type: 'EDIT_TRIP',
  id,
  updates
})

export const startEditTrip = (id, updates) => {
  return dispatch => {
    return db.collection('trips').doc(id).update(updates).then(() => {
      dispatch(editTrip(id, updates))
    })
  }
}

// REMOVE_TRIP
export const removeTrip = id => ({
  type: 'REMOVE_TRIP',
  id
})

export const startRemoveTrip = (id = '') => {
  return dispatch => {
    return db.collection('trips').doc(id).delete().then(() => {
      dispatch(removeTrip(id))
    })
  }
}

//SET_TRIPS
export const setTrips = trips => ({
  type: 'SET_TRIPS',
  trips
})

export const startSetTrips = () => {
  return dispatch => {
    return db.collection('trips').get().then(snapshot => {
      const trips = []

      snapshot.forEach(childSnapshot => {
        trips.push({
          id: childSnapshot.id,
          ...childSnapshot.data()
        })
      })

      dispatch(setTrips(trips))
    })
  }
}