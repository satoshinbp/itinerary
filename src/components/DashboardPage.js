import React from 'react';
import { Link } from 'react-router-dom';
import TripList from './TripList';
import TripListFilters from './TripListFilters';

const DashboardPage = () => (
  <div className="container">
    <div className="page-header">
      <Link className="btn-push" to="/create">Add Trip</Link>
      <TripListFilters />
    </div>
    <TripList />
  </div>
);

export default DashboardPage;