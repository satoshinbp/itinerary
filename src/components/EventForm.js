import React, { useState } from 'react'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'
import DateFnsUtils from '@date-io/date-fns'
import { TimePicker, MuiPickersUtilsProvider} from '@material-ui/pickers'
import moment from 'moment'

export default (props) => {
  const id = props.event ? props.event.id : props.id
  const tripId = props.event ? props.event.tripId : props.tripId
  const [title, setTitle] = useState(props.event ? props.event.title : '')
  const [date, setDate] = useState(props.event ? moment(props.event.date) : moment(props.date))
  const [startTime, setStartTime] = useState(props.event ? moment(props.event.startTime) : moment(props.date).hour(12).minutes(0))
  const [endTime, setEndTime] = useState(props.event ? moment(props.event.endTime) : moment(props.date).hour(12).minutes(0))
  const [location, setLocation] = useState(props.event ? props.event.location : '')
  const [note, setNote] = useState(props.event ? props.event.note : '')
  const [focused, setFocused] = useState(null)
  const [error, setError] = useState('')

  const onStartTimeChange = time => {
    setStartTime(moment(time))
    if (moment(time).isAfter(endTime)) {
      setEndTime(moment(time))
    }
  }
  const onEndTimeChange = time => {
    setEndTime(moment(time))
    if (moment(time).isBefore(startTime)) {
      setStartTime(moment(time))
    }
  }
  const onSubmit = e => {
    e.preventDefault()

    if (!title) {
      setError('Please provide title.')
    } else {
      setError('')
      props.onSubmit({
        id,
        tripId,
        title,
        date: date.valueOf(),
        startTime: startTime.valueOf(),
        endTime: endTime.valueOf(),
        location,
        note
      })
    }
  }

  return (
    <form className="form" onSubmit={onSubmit}>
      {error && <p className="form__error">{error}</p>}
      <input
        type="text"
        placeholder="title"
        autoFocus
        className="text-input"
        value={title}
        onChange={e => setTitle(e.target.value)}
      />
      <div className="date-time-picker">
        <SingleDatePicker
          date={date}
          onDateChange={date => setDate(moment(date).startOf('day'))}
          focused={focused}
          onFocusChange={({ focused }) => setFocused(focused)}
          isOutsideRange={() => false}
          displayFormat="YYYY/MM/DD"
        />
      </div>
      <div className="date-time-picker">
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={startTime} onChange={onStartTimeChange} label="Start Time" />
        </MuiPickersUtilsProvider>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={endTime} onChange={onEndTimeChange} label="End Time" />
        </MuiPickersUtilsProvider>
      </div>
      <input
        type="text"
        placeholder="location"
        className="text-input"
        value={location}
        onChange={e => setLocation(e.target.value)}
      />
      <textarea
        placeholder="Add a note for your event (optional)"
        className="textarea"
        value={note}
        onChange={e => setNote(e.target.value)}
      >
      </textarea>
      <button className="btn-push">Save Event</button>
    </form>
  )
}