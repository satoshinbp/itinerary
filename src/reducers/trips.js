const tripsReducerDefaultState = [];

export default (state = tripsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_TRIP':
      return [
        ...state,
        action.trip
      ];
    case 'EDIT_TRIP':
      return state.map((trip) => {
        if (trip.id === action.id) {
           return {
            ...trip,
            ...action.updates
           }
        } else {
          return trip;
        }
      });
    case 'REMOVE_TRIP':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};