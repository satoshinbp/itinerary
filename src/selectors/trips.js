import moment from 'moment';

export default (trips, { filter }) => {
  return trips.filter((trip) => {
    switch (filter) {
      case 'upcoming':
        return moment(trip.endDate).isSameOrAfter(moment(), 'day');
      case 'past':
        return moment(trip.endDate).isBefore(moment(), 'day');
      case 'all':
        return true;
    }
  }).sort((a, b) => {
    return a.startDate < b.startDate ? 1 : -1;
  });
};