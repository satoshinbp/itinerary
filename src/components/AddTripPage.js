import React from 'react';
import { connect } from 'react-redux';
import TripForm from './TripForm';
import { addTrip } from '../actions/trips';

const AddTripPage = (props) => (
  <div className="container">
    <TripForm
      onSubmit={(trip) => {
        props.dispatch(addTrip(trip));
        props.history.push('/');
      }} />
  </div>
);

export default connect()(AddTripPage);