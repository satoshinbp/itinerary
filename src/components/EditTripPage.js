import React from 'react';
import { connect } from 'react-redux';
import TripForm from './TripForm';
import { editTrip, removeTrip } from '../actions/trips';

const EditTripPage = (props) => {
  return (
    <div className="container">
      <TripForm
        trip={props.trip}
        onSubmit={(trip) => {
          props.dispatch(editTrip(props.trip.id, trip));
          props.history.push('/');
        }} />
      <button
        className="btn-push btn-push--secondary"
        onClick={() => {
          props.dispatch(removeTrip({ id: props.trip.id }));
          props.history.push('/');
        }}>Remove Expense</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  trip: state.trips.find((trip) => trip.id === props.match.params.id)
});

export default connect(mapStateToProps)(EditTripPage);
