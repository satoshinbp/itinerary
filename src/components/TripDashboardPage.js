import React, { useState } from 'react'
import Modal from 'react-modal'
import { connect } from 'react-redux'
import TripListButtons from './TripListButtons'
import TripList from './TripList'
import TripForm from './TripForm'
import { addTrip, editTrip } from '../actions/trips'

const TripDashboardPage = (props) => {
  const [tripId, setTripId] = useState(undefined)

  const onSubmit = (trip) => {
    if (props.trips.find((trip) => trip.id === tripId)) {
      props.editTrip(tripId, trip)
    } else {
      props.addTrip(trip)
    }
    setTripId(undefined)
  }

  return (
    <div className="container">
      <TripListButtons setTripId={setTripId} />
      <TripList setTripId={setTripId} />
      <Modal
        isOpen={!!tripId}
        contentLabel="Add/Edit Trip"
        onRequestClose={() => setTripId(undefined)}
        closeTimeoutMS={200}
        className="modal"
      >
        <TripForm
          onSubmit={onSubmit}
          tripId={tripId}
          trip={props.trips.find((trip) => trip.id === tripId)}
        />
      </Modal>
    </div>
  );
}

Modal.setAppElement('body')

const mapStateToProps = state => ({ trips: state.trips })

const mapDispatchToProps = dispatch => ({
  addTrip: trip => dispatch(addTrip(trip)),
  editTrip: (id, updates) => dispatch(editTrip(id, updates))
});

export default connect(mapStateToProps, mapDispatchToProps)(TripDashboardPage)