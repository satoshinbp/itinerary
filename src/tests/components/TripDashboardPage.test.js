import React from 'react'
import { shallow } from 'enzyme'
import trips from '../fixtures/trips'
import { TripDashboardPage } from '../../components/TripDashboardPage'

let addTrip, editTrip, setState, useStateSpy, wrapper
beforeEach(() => {
  addTrip = jest.fn()
  editTrip = jest.fn()
  setState = jest.fn()
  useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation(init => [init, setState])
  wrapper = shallow(<TripDashboardPage trips={trips} addTrip={addTrip} editTrip={editTrip} />)
})

test('should render TripDashboardPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should close Modal on request close', () => {
  wrapper.find('Modal').prop('onRequestClose')()
  expect(setState).toBeCalledWith(undefined)
})

// test for Modal open case to be implemented

// test('should submit TripForm on submit', () => {
//   wrapper.find('TripForm').prop('onSubmit')(trips[0])
//   expect(editTrip).toBeCalledWith(trips[0].id, trips[0])
//   expect(setState).toBeCalledWith(undefined)
// })