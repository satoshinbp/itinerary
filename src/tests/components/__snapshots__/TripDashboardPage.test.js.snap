// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TripDashboardPage correctly 1`] = `
<div
  className="container"
>
  <Connect(TripListButtons)
    setTripId={[MockFunction]}
  />
  <Connect(TripList)
    setTripId={[MockFunction]}
  />
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={200}
    contentLabel="Add/Edit Trip"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <_default
      onSubmit={[Function]}
    />
  </Modal>
</div>
`;
